localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,946 @@
+    }
+
+    if message_type == "Notification":
+        unsubscribe_url = create_unsubscribe_url(external_url, subscriber["SubscriptionArn"])
+        data["UnsubscribeURL"] = unsubscribe_url
+
+    elif message_type in ("UnsubscribeConfirmation", "SubscriptionConfirmation"):
+        data["Token"] = message_context.token
+        data["SubscribeURL"] = create_subscribe_url(
+            external_url, subscriber["TopicArn"], message_context.token
+        )
+
+    if message_context.subject:
+        data["Subject"] = message_context.subject
+
+    if message_context.message_attributes:
+        data["MessageAttributes"] = prepare_message_attributes(message_context.message_attributes)
+
+    return json.dumps(data)
+
+
+def prepare_message_attributes(
+    message_attributes: MessageAttributeMap,
+) -> Dict[str, Dict[str, str]]:
+    attributes = {}
+    if not message_attributes:
+        return attributes
+    # TODO: Number type is not supported for Lambda subscriptions, passed as String
+    #  do conversion here
+    for attr_name, attr in message_attributes.items():
+        data_type = attr["DataType"]
+        if data_type.startswith("Binary"):
+            # binary payload in base64 encoded by AWS, UTF-8 for JSON
+            # https://docs.aws.amazon.com/sns/latest/api/API_MessageAttributeValue.html
+            val = base64.b64encode(attr["BinaryValue"]).decode()
+        else:
+            val = attr.get("StringValue")
+
+        attributes[attr_name] = {
+            "Type": data_type,
+            "Value": val,
+        }
+    return attributes
+
+
+def is_raw_message_delivery(subscriber: SnsSubscription) -> bool:
+    return subscriber.get("RawMessageDelivery") in ("true", True, "True")
+
+
+def store_delivery_log(
+    message_context: SnsMessage, subscriber: SnsSubscription, success: bool, delivery: dict = None
+):
+    """Store the delivery logs in CloudWatch"""
+    log_group_name = subscriber.get("TopicArn", "").replace("arn:aws:", "").replace(":", "/")
+    log_stream_name = long_uid()
+    invocation_time = int(time.time() * 1000)
+
+    delivery = not_none_or(delivery, {})
+    delivery["deliveryId"] = (long_uid(),)
+    delivery["destination"] = (subscriber.get("Endpoint", ""),)
+    delivery["dwellTimeMs"] = 200
+    if not success:
+        delivery["attemps"] = 1
+
+    if (protocol := subscriber["Protocol"]) == "application":
+        protocol = get_platform_type_from_endpoint_arn(subscriber["Endpoint"])
+
+    message = message_context.message_content(protocol)
+    delivery_log = {
+        "notification": {
+            "messageMD5Sum": md5(message),
+            "messageId": message_context.message_id,
+            "topicArn": subscriber.get("TopicArn"),
+            "timestamp": datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f%z"),
+        },
+        "delivery": delivery,
+        "status": "SUCCESS" if success else "FAILURE",
+    }
+
+    log_output = json.dumps(json_safe(delivery_log))
```

Comment:
is `json_safe` necessary here? is there a situation where there are non-serializable attributes in `delivery_log`?

Reply:
I'm actually not sure, I don't think so. I reused the utility method, I'll run the tests without it, but from a quick look, it does not seem like it.